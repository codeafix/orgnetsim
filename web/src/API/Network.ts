type Network = {
    links: Array<Link>
    nodes: Array<AgentState>;
    maxColors: number;
}

type Link = {
    source: string;
    target: string;
    strength: number;
    length: number;
}

type AgentState = {
    id: string;
    color: number;
    susceptability: number;
    influence: number;
    contrariness: number;
    change: number;
    type: string;
    fx: number;
    fy: number;
}
