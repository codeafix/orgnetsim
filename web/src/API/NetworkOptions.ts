type NetworkOptions = {
    linkTeamPeers: boolean;
    linkedTeamList: Array<string>;
    evangelistList: Array<string>;
    loneEvangelist: Array<string>;
    initColors: Array<number>;
    maxColors: number;
    agentsWithMemory: boolean;
}