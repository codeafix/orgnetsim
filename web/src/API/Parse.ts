type ParseOptions = {
    identifier: number;
    parent: number;
    regex: any;
    delimiter: string;
    payload: string;
}

export type { ParseOptions };