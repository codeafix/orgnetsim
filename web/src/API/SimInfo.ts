type SimInfo = {
    id: string;
    name: string;
    description: string;
    steps: Array<string>;
    options: NetworkOptions;
}