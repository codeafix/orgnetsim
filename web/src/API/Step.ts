import { Results } from './Results';

type Step = {
    id: string;
    parent: string;
    results: Results;
}

type RunSpec = {
    steps: number;
    iterations: number;
}

export type { Step, RunSpec };