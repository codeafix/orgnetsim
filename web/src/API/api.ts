const API = {
    rootPath: "http://localhost:8080",
    emptySimList: {simulations:[],notes:""},
    sims: async function() {
        const response = await fetch(this.rootPath+"/api/simulation", {
        "method": "GET",
        "headers": {}
        }).catch(err => { console.log(err); 
        });
        return response.json();
    },
    get: async function(id){
        const response = await fetch(this.rootPath+"/api/simulation/"+id, {
            "method": "GET",
            "headers": {}
            }).catch(err => { console.log(err); 
        });
        return response.json();
    },
    getStep: async function(path){
        const response = await fetch(this.rootPath+path, {
            "method": "GET",
            "headers": {}
            }).catch(err => { console.log(err); 
        });
        return response.json();
    },
    getSteps: async function(sim){
        const response = await fetch(this.rootPath+"/api/simulation/"+sim.id+"/step", {
            "method": "GET",
            "headers": {}
            }).catch(err => { console.log(err); 
        });
        return response.json();
    },
    getResults: async function(sim){
        const response = await fetch(this.rootPath+"/api/simulation/"+sim.id+"/results", {
            "method": "GET",
            "headers": {}
            }).catch(err => { console.log(err); 
        });
        
        return response.json();
    },
    getResultsCsv: async function(sim){
        var fn = "results.csv";
        const data = await fetch(this.rootPath+"/api/simulation/"+sim.id+"/results", {
            "method": "GET",
            "headers": {
                "Content-Type": "text/csv"
            }
            }).then(response => {
                const contentDisp = response.headers.get('Content-Disposition');
                const regExpFilename = /filename="(?<filename>[^"]*)"/;
                fn = regExpFilename.exec(contentDisp)?.groups?.filename ?? fn;
                return response.blob();
            }).then(blob => {
                return {
                    filename: fn,
                    blob: blob,
                };
            }).catch(err => { console.log(err); 
            });
        return data;
    },
    update: async function(sim){
        const response = await fetch(this.rootPath+"/api/simulation/"+sim.id, {
            "method": "PUT",
            "headers": {
                'Content-Type': 'application/json'
            },
            "body": JSON.stringify(sim),
            })
            .catch(err => { console.log(err); 
            });
        return response.json();
    },
    updateStep: async function(step){
        const response = await fetch(this.rootPath+"/api/simulation/"+step.parent+"/step/"+step.id, {
            "method": "PUT",
            "headers": {
                'Content-Type': 'application/json'
            },
            "body": JSON.stringify(step),
            })
            .catch(err => { console.log(err); 
            });
        return response.json();
    },
    runsim: async function(sim, spec){
        const response = await fetch(this.rootPath+"/api/simulation/"+sim.id+"/run", {
            "method": "POST",
            "headers": {
                'Content-Type': 'application/json'
            },
            "body": JSON.stringify(spec),
            })
            .catch(err => { console.log(err); 
            });
            return response.json();
    },
    parse: async function(sim, pdata){
        const response = await fetch("http://localhost:8080/api/simulation/"+sim.id+"/parse", {
            "method": "POST",
            "headers": {
                'Content-Type': 'application/json'
            },
            "body": JSON.stringify(pdata),
            }).catch(err => { console.log(err); 
            });
            return response.json();
    },
    add: async function(name, description){
        var sim = {name:name,description:description};
        const response = await fetch(this.rootPath+"/api/simulation", {
            "method": "POST",
            "headers": {
                'Content-Type': 'application/json'
            },
            "body": JSON.stringify(sim),
            })
            .catch(err => { console.log(err); 
            });
        return response.json();
    },
    delete: async function(id){
        await fetch(this.rootPath+"/api/simulation/"+id, {
            "method": "DELETE",
            "headers": {}
            })
            .catch(err => { console.log(err); 
            });
    },
}

export default API