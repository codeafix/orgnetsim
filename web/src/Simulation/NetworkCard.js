import React, { useState, useEffect } from 'react';
import {Card, Form, Modal, FormControl, Button} from 'react-bootstrap'
import API from '../api';
import NetworkGraph from './NetworkGraph';

const NetworkCard = (props) => {
    const [idcol, setidcol] = useState(0);
    const [pcol, setpcol] = useState(2);
    const [delim, setdelim] = useState(",");
    const [filetoupload, setfiletoupload] = useState();
    const [showimpmodal, setshowimpmodal] = useState(false);
    const [hasstep, sethasstep] = useState(false);

    useEffect(() => {
        sethasstep((props.steps || []).length > 0)
    },[props.sim, props.steps]);

    const handleimpclose = () => {
        setshowimpmodal(false);
        setidcol(0);
        setpcol(2);
        setdelim(",");
        setfiletoupload();
    };
    
    const handleimport = () => {
        const fr = new FileReader();

        fr.readAsText(filetoupload);
        fr.onload = function() {
            const base64data = btoa(fr.result);
        
            const pdata = {
                "identifier": idcol,
                "parent": pcol,
                "regex": {
                    "0":"^\\d+_(.*)$",
                    "2":"^\\d+_(.*)$",
                    "3":"\\S+"
                  },
                "delimiter": delim,
                "Payload": base64data
            };
            API.parse(props.sim, pdata).then(response => {
                props.readsim(response.parent);
            });
        };
        handleimpclose();
    };
    
    return(
        <Card className="mb-2 mx-n2">
            <Card.Header><Card.Title>Network<Button size="sm" className="btn btn-primary float-right" onClick={() => setshowimpmodal(true)} disabled={hasstep}>Import</Button></Card.Title></Card.Header>
            <Card.Body>
                <NetworkGraph sim={props.sim} steps={props.steps}/>
            </Card.Body>
            <Modal
                show={showimpmodal}
                onHide={handleimpclose}
                backdrop="static"
                keyboard={false}
            >
                <Modal.Header closeButton>
                    <Modal.Title>Import Network</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    <Form.Group controlId="form-file">
                        <Form.File label="Select File" onChange={(e) => {
                            if(e.target.files.length) setfiletoupload(e.target.files[0]);
                        }}/>
                    </Form.Group>
                    <Form.Group controlId="form-identifier">
                        <Form.Label>Identifier column</Form.Label>
                        <Form.Control type="number" value={idcol} onChange={e => setidcol(e.target.valueAsNumber)}/>
                        <Form.Text className="text-muted">
                            The column that has the unique identifier in it
                        </Form.Text>
                    </Form.Group>
                    <Form.Group controlId="form-parent">
                        <Form.Label>Parent Identifier Column</Form.Label>
                        <Form.Control type="number" value={pcol} onChange={e => setpcol(e.target.valueAsNumber)}/>
                        <Form.Text className="text-muted">
                            The column that has a parent identifier in it to show hierarchy in the network
                        </Form.Text>
                    </Form.Group>
                    <Form.Group className="mb-3" controlId="form-delimiter">
                        <Form.Label>Delimiter</Form.Label>
                        <FormControl type="string" value={delim} onChange={e => setdelim(e.target.value)}/>
                        <Form.Text className="text-muted">
                            The delimiter used to separate columns in rows of the imported file
                        </Form.Text>
                    </Form.Group>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="success" onClick={handleimport}>Save</Button>
                    <Button variant="secondary" onClick={handleimpclose}>Cancel</Button>
                </Modal.Footer>
            </Modal>
        </Card>
    )
}

export default NetworkCard