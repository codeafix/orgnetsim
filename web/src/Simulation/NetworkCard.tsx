import React, { useState } from 'react';
import {Card, Button} from 'react-bootstrap'
import NetworkGraph from './NetworkGraph';
import ParseSettings from './ParseSettings';
import { SimInfo } from '../API/SimInfo';
import { Step } from '../API/Step';

type NetworkCardProps = {
    sim:SimInfo;
    steps:Array<Step>;
    readsim(id:string): void;
}

const NetworkCard = (props:NetworkCardProps) => {
    const [showimpmodal, setshowimpmodal] = useState<boolean>(false);
    
    return(
        <Card className="mb-2 mx-n2">
            <Card.Header><Card.Title>Network<Button size="sm" className="btn btn-primary float-right" onClick={() => setshowimpmodal(true)}>Import</Button></Card.Title></Card.Header>
            <Card.Body>
                <NetworkGraph sim={props.sim} steps={props.steps}/>
            </Card.Body>
            <ParseSettings sim={props.sim} steps={props.steps} readsim={props.readsim} showimportmodal={showimpmodal} onclose={() => setshowimpmodal(false)}/>
        </Card>
    )
}

export default NetworkCard