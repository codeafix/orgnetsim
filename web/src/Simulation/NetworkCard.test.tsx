import React from 'react';
import {render} from '@testing-library/react';
import '../../node_modules/bootstrap/dist/css/bootstrap.css';
import '../index.css';
import { test, expect, vi } from 'vitest'
import { act } from 'react-dom/test-utils';
import NetworkCard from './NetworkCard';
import { SimInfo } from '../API/SimInfo';
import { Step } from '../API/Step';

const sim:SimInfo = {"id":"27f06fe2-6e82-44b0-af4a-6975d169ff48","name":"test","description":"","steps":["/api/simulation/27f06fe2-6e82-44b0-af4a-6975d169ff48/step/72e1e5cb-3f31-4afd-818f-2293076547f7","/api/simulation/27f06fe2-6e82-44b0-af4a-6975d169ff48/step/f62bd8e5-2027-4fca-9e02-92c6ae6468ac"],"options":{"linkTeamPeers":true,"linkedTeamList":[],"evangelistList":[],"loneEvangelist":[],"initColors":[0],"maxColors":2,"agentsWithMemory":false}};
const steps:Array<Step> = [{"results":{"iterations":0,"colors":[[6,0]],"conversations":[0]},"id":"72e1e5cb-3f31-4afd-818f-2293076547f7","parent":"27f06fe2-6e82-44b0-af4a-6975d169ff48"},{"results":{"iterations":100,"colors":[[4,2],[6,0],[5,1],[4,2],[4,2],[4,2],[5,1],[5,1],[5,1],[5,1],[6,0],[4,2],[4,2],[4,2],[4,2],[5,1],[4,2],[4,2],[5,1],[5,1],[6,0],[5,1],[6,0],[6,0],[6,0],[6,0],[6,0],[5,1],[5,1],[5,1],[6,0],[4,2],[5,1],[4,2],[5,1],[5,1],[5,1],[6,0],[5,1],[4,2],[4,2],[6,0],[5,1],[5,1],[5,1],[6,0],[5,1],[5,1],[5,1],[5,1],[4,2],[5,1],[5,1],[5,1],[4,2],[5,1],[6,0],[6,0],[6,0],[6,0],[5,1],[5,1],[5,1],[5,1],[6,0],[4,2],[5,1],[6,0],[4,2],[5,1],[5,1],[5,1],[5,1],[4,2],[5,1],[4,2],[5,1],[4,2],[5,1],[6,0],[5,1],[5,1],[6,0],[6,0],[5,1],[5,1],[6,0],[4,2],[4,2],[5,1],[3,3],[3,3],[4,2],[3,3],[4,2],[4,2],[4,2],[6,0],[4,2],[4,2]],"conversations":[6,6,6,6,5,6,6,6,6,6,6,6,6,6,6,5,6,6,6,6,6,6,6,6,6,6,6,5,6,6,6,6,6,6,6,6,6,5,6,6,6,6,6,6,6,5,6,6,6,6,6,5,6,6,6,6,6,6,6,5,6,6,6,6,6,6,6,6,6,6,6,6,5,6,6,6,6,6,6,6,5,6,6,6,6,6,5,6,5,6,6,6,6,5,6,6,6,6,6,6]},"id":"f62bd8e5-2027-4fca-9e02-92c6ae6468ac","parent":"27f06fe2-6e82-44b0-af4a-6975d169ff48"}]

vi.mock('../API/api');

test('renders without crashing', async () => {
    var result:any;
    await act(async () => {
        result = render(
            <NetworkCard sim={sim} steps={steps} readsim={function (id: string): void {
                throw new Error('Function not implemented.');
            } }/>
        );
    });
    expect(result).toBeDefined();
    expect(result.asFragment()).toMatchSnapshot();
});
