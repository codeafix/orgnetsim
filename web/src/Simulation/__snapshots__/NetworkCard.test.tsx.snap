// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="mb-2 mx-n2 card"
  >
    <div
      class="card-header"
    >
      <div
        class="card-title h5"
      >
        Network
        <button
          class="btn btn-primary float-right btn btn-primary btn-sm"
          type="button"
        >
          Import
        </button>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        id="graph-container"
      >
        <svg
          class="mb-3"
          height="0"
          preserveAspectRatio="xMinYMid"
          viewBox="0 0 0 0"
          width="0"
        >
          <g
            class="grpParent"
          >
            <g
              class="links"
            >
              <line
                style="stroke: LightGray; stroke-width: 0.175;"
                x1="512.0238945054554"
                x2="528.3743371463426"
                y1="383.2777626590721"
                y2="403.6228601629286"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.175;"
                x1="512.0238945054554"
                x2="486.83099078103857"
                y1="383.2777626590721"
                y2="378.383316333952"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.2;"
                x1="512.0238945054554"
                x2="536.0980539104477"
                y1="383.2777626590721"
                y2="374.5974462319489"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.225;"
                x1="512.0238945054554"
                x2="498.7319761723783"
                y1="383.2777626590721"
                y2="406.0009971245655"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.205;"
                x1="512.0238945054554"
                x2="509.9407199938294"
                y1="383.2777626590721"
                y2="358.1175533021693"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.235;"
                x1="528.3743371463426"
                x2="486.83099078103857"
                y1="403.6228601629286"
                y2="378.383316333952"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.225;"
                x1="528.3743371463426"
                x2="536.0980539104477"
                y1="403.6228601629286"
                y2="374.5974462319489"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.18;"
                x1="528.3743371463426"
                x2="498.7319761723783"
                y1="403.6228601629286"
                y2="406.0009971245655"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.185;"
                x1="528.3743371463426"
                x2="509.9407199938294"
                y1="403.6228601629286"
                y2="358.1175533021693"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.17;"
                x1="486.83099078103857"
                x2="536.0980539104477"
                y1="378.383316333952"
                y2="374.5974462319489"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.195;"
                x1="486.83099078103857"
                x2="498.7319761723783"
                y1="378.383316333952"
                y2="406.0009971245655"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.185;"
                x1="486.83099078103857"
                x2="509.9407199938294"
                y1="378.383316333952"
                y2="358.1175533021693"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.17;"
                x1="536.0980539104477"
                x2="498.7319761723783"
                y1="374.5974462319489"
                y2="406.0009971245655"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.225;"
                x1="536.0980539104477"
                x2="509.9407199938294"
                y1="374.5974462319489"
                y2="358.1175533021693"
              />
              <line
                style="stroke: LightGray; stroke-width: 0.19;"
                x1="498.7319761723783"
                x2="509.9407199938294"
                y1="406.0009971245655"
                y2="358.1175533021693"
              />
            </g>
            <g
              class="nodes"
            >
              <circle
                cx="512.0238945054554"
                cy="383.2777626590721"
                r="5.1125"
                style="fill: Blue;"
              >
                <title>
                  Lead
                </title>
              </circle>
              <circle
                cx="498.7319761723783"
                cy="406.0009971245655"
                r="5.0325"
                style="fill: Blue;"
              >
                <title>
                  Dev 1
                </title>
              </circle>
              <circle
                cx="509.9407199938294"
                cy="358.1175533021693"
                r="5"
                style="fill: SlateGray;"
              >
                <title>
                  Dev 2
                </title>
              </circle>
              <circle
                cx="528.3743371463426"
                cy="403.6228601629286"
                r="5"
                style="fill: SlateGray;"
              >
                <title>
                  Dev 3
                </title>
              </circle>
              <circle
                cx="486.83099078103857"
                cy="378.383316333952"
                r="5.045"
                style="fill: SlateGray;"
              >
                <title>
                  Dev 4
                </title>
              </circle>
              <circle
                cx="536.0980539104477"
                cy="374.5974462319489"
                r="5"
                style="fill: SlateGray;"
              >
                <title>
                  Dev 5
                </title>
              </circle>
            </g>
          </g>
        </svg>
        <button
          class="btn btn-primary float-right btn btn-primary btn-sm"
          type="button"
        >
          Play
        </button>
        <button
          class="btn btn-primary float-right btn btn-primary btn-sm"
          disabled=""
          type="button"
        >
          Layout
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
